com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```diff
@@ -0,0 +1,769 @@
+          && linkBundle.getLinkedWorkItem().getId().equals(linkStruct.getLinkedItem().getId())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Binary search to cut down the search space since the list is ordered in ascending order* */
+  private int searchIndexWorkItemHistory(IPObjectList<IWorkItem> workItemHistory) {
+
+    if (workItemHistory == null) return -1;
+
+    Integer lastUpdateInt = Integer.valueOf(lastUpdate);
+    Integer endRevisionInt = Integer.valueOf(endRevision);
+    Integer lastItemRevision =
+        Integer.valueOf(workItemHistory.get(workItemHistory.size() - 1).getRevision());
+    Integer firstItemVersion = Integer.valueOf(workItemHistory.get(0).getRevision());
+
+    // Short circuit: don't need to binarysearch if you're looking for all history
+    if (lastUpdateInt <= 0) return 0;
+
+    // Short circuit: don't need to search if lastUpdate points to the last version of the item
+    if (lastItemRevision == lastUpdateInt) return workItemHistory.size() - 1;
+
+    // Short circuit: don't search if all changes are before lastUpdate or after endRevision
+    if (lastItemRevision < lastUpdateInt || firstItemVersion > endRevisionInt) return -1;
+
+    int left = 0;
+    int right = workItemHistory.size() - 1;
+    int index = -1;
+
+    while (left <= right) {
+      int mid = (left + right) / 2;
+      if (Integer.valueOf(workItemHistory.get(mid).getRevision()) < lastUpdateInt) {
+        left = mid + 1;
+      } else if (Integer.valueOf(workItemHistory.get(mid).getRevision()) == lastUpdateInt) {
+        // When all changes came after the lastUpdate and endRevision
+        if (firstItemVersion == Integer.valueOf(workItemHistory.get(mid).getRevision())) {
+          return -1;
+        }
+        return mid;
+      } else {
+        index = mid;
+        right = mid - 1;
+      }
+    }
+
+    // When the item changed before lastUpdate and afterEndRevision
+    if (Integer.valueOf(workItemHistory.get(index).getRevision()) < lastUpdateInt
+        || Integer.valueOf(workItemHistory.get(index).getRevision()) > endRevisionInt) {
+      return -1;
+    }
+
+    return (index == 0 ? index : index - 1);
+  }
+
+  /*
+   * This method validates these required request attributes.
+   * Split into three separate helper methods, one for each param.
+   *  */
+  private boolean validateParameters(String projectId, String space, String doc) {
+    // Needs to be executed in this order. Space validation only runs after projectId is validated.
+    // DocId is validated only if projectId and SpaceId are validated.
+    return validateProjectId(projectId)
+        && validateSpaceId(projectId, space)
+        && validateDocumentId(projectId, space, doc);
+  }
+
+  private boolean validateProjectId(String projectId) {
+    try {
+      IProject projObj = trackerService.getProjectsService().getProject(projectId);
+
+      logger.info("[Teamscale Polarion Plugin] Attempting to read projectID: " + projObj.getId());
```

Comment:
Yeah, good point. I wasn't sure if this logger will be the final solution. I haven't found documentation around logging in Polarion. I just grabbed that Logger that I found in their API. Then Polarion dumps it in the internal log folder in the Polarion installation. I guess for now we can stick with it for now, but in a future version it would be good to have our own log configuration.

Done.